cpp/src/iterator/iterator.cuh

```diff
@@ -0,0 +1,382 @@
+
+#ifndef CUDF_ITERATOR_CUH
+#define CUDF_ITERATOR_CUH
+
+#include <cudf/cudf.h>
+#include <iterator/transform_unary_functions.cuh>
+
+#include <bitmask/bit_mask.cuh>         // need for bit_mask::bit_mask_t
+#include <utilities/cudf_utils.h>       // need for CUDA_HOST_DEVICE_CALLABLE
+#include <utilities/error_utils.hpp>
+#include <utilities/type_dispatcher.hpp>
+
+#include <thrust/iterator/iterator_adaptor.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/transform_iterator.h>
+#include <thrust/pair.h>
+#include <type_traits>
+
+namespace cudf
+{
+
+/** -------------------------------------------------------------------------*
+ * @brief value accessor with/without null bitmask
+ * A unary function returns scalar value at `id`.
+ * `operator() (gdf_index_type id)` computes
+ * `data` value and valid flag at `id`
+ *
+ * If has_nulls = false, the data is always `static_cast<ResultType>(data[id])`
+ * If has_nulls = true,  the data value = (is_valid(id))?
+ *      static_cast<ResultType>(data[id]) : identity;
+ *
+ * @tparam  ElementType cudf data type of input element array
+ * @tparam  ResultType  cudf data type of output and `identity` value
+ *                    which is used when null bitmaps flag is false.
+ * @tparam  has_nulls if false, this struct holds only data array.
+ *                    else, this struct holds data array and
+ *                    bitmask array and identity value
+ * -------------------------------------------------------------------------**/
+template <typename ElementType, typename ResultType, bool has_nulls>
+struct value_accessor;
+
+/** -------------------------------------------------------------------------*
+ * @overload value_accessor<ElementType, ResultType, true>
+ * @brief specialization for columns that contain null values
+ * -------------------------------------------------------------------------**/
+template <typename ElementType, typename ResultType>
+struct value_accessor<ElementType, ResultType, true>
+{
+  ElementType const* elements{};          ///< pointer of cudf data array
+  bit_mask::bit_mask_t const* bitmask{};  ///< pointer of cudf bitmask (null) array
+  ResultType const identity{};            ///< identity value used when the validity is false
+
+/** -------------------------------------------------------------------------*
+ * @brief constructor
+ * @param[in] e pointer of cudf data array
+ * @param[in] b pointer of cudf bitmask (null) array
+ * @param[in] i identity value used when the validity is false at operator()
+ * -------------------------------------------------------------------------**/
+  value_accessor(ElementType const* e, bit_mask::bit_mask_t const* b, ResultType i)
+    : elements{e}, bitmask{b}, identity{i}
+  {
+#if  !defined(__CUDA_ARCH__)
+    // verify valid is non-null, otherwise, is_valid() will crash
+    CUDF_EXPECTS(b != nullptr, "non-null bit mask is required");
+#endif
+  }
+
+  CUDA_HOST_DEVICE_CALLABLE
```

Comment:
Should this be host-callable? I believe the bitmask and the elements are always device pointers, aren't they?

Reply:
At first, I add thrust host test case for debugging and make sure if the iterator works well. It's already confirmed, then __device__ should be enough.
I've changed the value_accessor and pair_accessor intp `CUDA_DEVICE_CALLABLE`.

But for functor in transform_unary_functions.cuh, I still use `CUDA_HOST_DEVICE_CALLABLE` since there may be possibility to use them at host code.